### PR TITLE
phpExtensions.grpc: use shared system grpc

### PR DESCRIPTION
**Pull request by shyim**
## Description of changes

This allows us to use the system grpc instead of compiling grpc with his libraries again and then use it together with the PHP extension.

This reduces the build time from **~20 minutes** to just 5 seconds. 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by drupol** [comment]
This is beautiful ! I'll let @jtojnar or @etu have their word on this !

File: pkgs/development/php-packages/grpc/default.nix
Comment (jtojnar) [inline pkgs/development/php-packages/grpc/default.nix:36]: Ideally, we would use pkg-config like other extensions since PHP 7.4/8. That should make the patch more upstreamable.

Comment (shyim) [reply pkgs/development/php-packages/grpc/default.nix:36]: I am not sure why they currently pecl zip have extension + grpc + third party libraries included and compiles everything together. I think they do that to have the same version of grpc library to php extension version, thats why I reuse the source of grpc.

I asked for pkg-config in my PR: https://github.com/grpc/grpc/pull/35404#issuecomment-1871022142

Comment (shyim) [reply pkgs/development/php-packages/grpc/default.nix]: I updated the patch to use pkg-config

File: pkgs/development/php-packages/grpc/default.nix
Comment (drupol) [inline pkgs/development/php-packages/grpc/default.nix:10]: Wouldn't it be a good idea to keep `buildPecl` and override the `src` attribute?

File: pkgs/development/php-packages/grpc/default.nix
Comment (jtojnar) [inline pkgs/development/php-packages/grpc/default.nix:24]: I would expect grpc to be a runtime dependency.

File: pkgs/development/php-packages/grpc/default.nix
Comment (shyim) [reply pkgs/development/php-packages/grpc/default.nix]: Ahh yea, we can still use it :)

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83

**Review by drupol — APPROVED** [review]
LGTM, I'll wait for @jtojnar for the merge.

File: pkgs/development/php-packages/grpc/default.nix
Comment (jtojnar) [inline pkgs/development/php-packages/grpc/default.nix:22]: Does `checkFlags` not work for `NO_INTERACTION`? If not, we should run `pre`/`postCheck` hooks here. And drop `checkTarget` since that is redundant.

Comment (jtojnar) [reply pkgs/development/php-packages/grpc/default.nix:22]: Actually, looks like `buildPecl` already sets `checkPhase` like this, so all of these can be dropped.

`buildPecl` should probably be adjusted as suggested above. But that is for a separate PR.

File: pkgs/development/php-packages/grpc/default.nix
Comment (jtojnar) [inline pkgs/development/php-packages/grpc/default.nix:12]: Slightly less fragile:

```suggestion
  sourceRoot = "${grpc.src.name}/src/php/ext/grpc";
```

Though ideally, we would have a separate variable for entering a directory after entering a source root.

File: pkgs/development/php-packages/grpc/default.nix
Comment (jtojnar) [inline pkgs/development/php-packages/grpc/default.nix:22]: This is nor needed.

Comment (drupol) [reply pkgs/development/php-packages/grpc/default.nix:22]: But this is out of scope of this PR, no ?

Comment (shyim) [reply pkgs/development/php-packages/grpc/default.nix]: done :)

**Comment by drupol** [comment]
@GrahamcOfBorg build php81.extensions.grpc php82.extensions.grpc php83.extensions.grpc

**Review by drupol — APPROVED** [review]

File: pkgs/development/php-packages/grpc/default.nix
Comment (drupol) [reply pkgs/development/php-packages/grpc/default.nix]: Sorry for my previous feedback, my brain farted.

**Comment by drupol** [comment]
@jtojnar good for you ?

**Comment by drupol** [comment]
@shyim There's a build issue, can you have a look?

**Comment by shyim** [comment]
@GrahamcOfBorg build php81.extensions.grpc php82.extensions.grpc php83.extensions.grpc

**Comment by shyim** [comment]
I have no idea, it works locally for me. I am even on `aarch64-darwin`

**Comment by drupol** [comment]
@GrahamcOfBorg build phpExtensions.grpc

**Pull request merged**
